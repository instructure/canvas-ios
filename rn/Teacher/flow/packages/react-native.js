//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

declare var console: typeof console & { disableYellowBox: boolean }

export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}

export type KeyboardEventData = {
  endCoordinates: {
    width: number,
    height: number,
    screenX: number,
    screenY: number,
  },
}
