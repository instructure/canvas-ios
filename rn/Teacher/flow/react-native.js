//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow
/* eslint-disable */

// define basic style prop types
// the internal ones are the best we have
import {
  type ____ViewStyleProp_Internal,
  type ____TextStyleProp_Internal,
  type ____ImageStyleProp_Internal,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes'

declare type ViewStyleProp = ____ViewStyleProp_Internal
declare type TextStyleProp = ____TextStyleProp_Internal
declare type ImageStyleProp = ____ImageStyleProp_Internal
