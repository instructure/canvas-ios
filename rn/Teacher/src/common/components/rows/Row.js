//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import React, { Component } from 'react'
import {
  View,
  Image,
  TouchableHighlight,
} from 'react-native'
import { colors, createStyleSheet } from '../../stylesheet'

import DisclosureIndicator from '../DisclosureIndicator'
import { Text } from '../../text'

export type RowProps = {
  title: string,
  subtitle?: ?string,
  image?: { uri: string } | number,
  imageTint?: string,
  imageSize?: { height: number, width: number }, // Defaults to 20 x 20 if not supplied
  disclosureIndicator?: boolean,
  border?: 'top' | 'bottom' | 'both',
  onPress?: Function,
  testID?: string,
  children?: React$Node,
  renderImage?: Function,
  identifier?: any, // Passed in as the first parameter to the onPress callback
  accessories?: any,
  accessibilityLabel?: ?string,
  accessibilityRole?: string,
  accessibilityState?: any,
  titleProps?: { ellipsizeMode?: string, numberOfLines?: number, accessibilityLabel?: string },
  selected?: ?boolean,
  titleStyles?: any,
  subtitleStyles?: any,
}

export default class Row extends Component<RowProps> {
  onPress = () => {
    if (this.props.onPress) {
      this.props.onPress(this.props.identifier)
    }
  }

  render () {
    const imageSize = this.props.imageSize || { height: 20, width: 20 }
    const title = this.props.title
    const testID = this.props.testID || 'row.undefined-cell'
    const hasIcon = this.props.image || this.props.renderImage

    let topBorder
    let bottomBorder

    if (this.props.border === 'top' || this.props.border === 'both') {
      topBorder = style.topHairline
    }

    if (this.props.border === 'bottom' || this.props.border === 'both') {
      bottomBorder = style.bottomHairline
    }

    let accessibilityRole = this.props.accessibilityRole

    if (accessibilityRole === undefined && this.props.onPress) {
      accessibilityRole = 'button'
    }

    let traits: {[string]: any} = {
      accessibilityRole,
      accessibilityLabel: this.props.accessibilityLabel,
      accessibilityState: this.props.accessibilityState,
    }

    if (this.props.accessible !== undefined) {
      traits['accessible'] = this.props.accessible
    }

    let underlayProps = {}
    let backgroundColor = colors.backgroundLightest
    if (this.props.selected) {
      underlayProps.underlayColor = colors.backgroundLight
      backgroundColor = colors.backgroundLight
    }

    // This broke highlighting basically everywhere, so it needs to be disabled.
    // if (this.props.selected !== undefined) {
    //   // This row uses selected state so dont ever show active state
    //   underlayProps.activeOpacity = 1
    // }

    const titleStyles = [style.title, this.props.titleStyles].filter(Boolean)
    const subtitleStyles = [style.subtitle, this.props.subtitleStyles].filter(Boolean)
    const RowView = this.props.onPress ? TouchableHighlight : View
    let titleA11y = this.props.titleProps?.accessibilityLabel

    return (<RowView style={[topBorder, bottomBorder]} { ...traits } onPress={this.onPress} testID={this.props.testID} {...underlayProps}>
      <View style={[style.container, { backgroundColor }, this.props.style]}>
        { this.props.renderImage && this.props.renderImage() }
        { this.props.image && <Image style={[style.image, { tintColor: this.props.imageTint, height: imageSize.height, width: imageSize.width }]} source={this.props.image} /> }
        <View style={[style.titlesContainer, { marginLeft: hasIcon ? 12 : 0 }]}>
          { Boolean(title) &&
            <Text
              style={titleStyles}
              ellipsizeMode={this.props.titleProps?.ellipsizeMode || 'tail'}
              numberOfLines={this.props.titleProps?.numberOfLines || 0}
              {... (titleA11y ? { accessibilityLabel: titleA11y } : {})}
            >
              {title}
            </Text>
          }
          { Boolean(this.props.subtitle) && <Text style={subtitleStyles} testID={`${testID}-subtitle-lbl`}>{this.props.subtitle}</Text> }
          { this.props.children }
        </View>
        { Boolean(this.props.accessories || this.props.disclosureIndicator) &&
          <View style={style.accessoryContainer}>
            <View style={style.accessories}>
              { this.props.accessories }
              { this.props.disclosureIndicator && <DisclosureIndicator /> }
            </View>
          </View>
        }
      </View>
    </RowView>)
  }
}

const style = createStyleSheet((colors, vars) => ({
  container: {
    flex: 1,
    flexDirection: 'row',
    minHeight: 54,
    alignItems: 'center',
    paddingTop: Math.floor(vars.padding / 1.25),
    paddingBottom: Math.floor(vars.padding / 1.25),
    paddingLeft: vars.padding,
    paddingRight: vars.padding,
    backgroundColor: colors.backgroundLightest,
  },
  titlesContainer: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'center',
  },
  accessoryContainer: {
    flex: 0,
  },
  accessories: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-end',
  },
  topHairline: {
    borderTopWidth: vars.hairlineWidth,
    borderTopColor: colors.borderMedium,
  },
  bottomHairline: {
    borderBottomWidth: vars.hairlineWidth,
    borderBottomColor: colors.borderMedium,
  },
  title: {
    color: colors.textDarkest,
    fontWeight: '600',
  },
  subtitle: {
    color: colors.textDark,
    fontSize: 14,
    marginTop: 2,
  },
  image: {
    resizeMode: 'contain',
  },
}))
