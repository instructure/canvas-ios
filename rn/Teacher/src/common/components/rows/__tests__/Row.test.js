//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import 'react-native'
import React from 'react'
import Row from '../Row'
import Images from '../../../../../src/images'

import renderer from 'react-test-renderer'

test('Render the base row', () => {
  const onPress = jest.fn()
  let aRow = renderer.create(
    <Row
      title='Title of the row'
      subtitle='Subtitle'
      image={Images.invisible}
      imageTint='#fff'
      imageSize={{ height: 10, width: 10 }}
      disclosureIndicator={true}
      height={44}
      border='both'
      onPress={onPress} />
  )
  expect(aRow.toJSON()).toMatchSnapshot()
})
