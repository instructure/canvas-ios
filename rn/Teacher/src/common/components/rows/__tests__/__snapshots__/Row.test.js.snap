// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render the base row 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "borderTopColor": "#C7CDD1",
        "borderTopWidth": 0.5,
      },
      Object {
        "borderBottomColor": "#C7CDD1",
        "borderBottomWidth": 0.5,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "flex": 1,
          "flexDirection": "row",
          "minHeight": 54,
          "paddingBottom": 12,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 12,
        },
        Object {
          "backgroundColor": "#FFFFFF",
        },
        undefined,
      ]
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../src/images/invisible.png",
        }
      }
      style={
        Array [
          Object {
            "resizeMode": "contain",
          },
          Object {
            "height": 10,
            "tintColor": "#fff",
            "width": 10,
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "center",
          },
          Object {
            "marginLeft": 12,
          },
        ]
      }
    >
      <Text
        ellipsizeMode="tail"
        numberOfLines={0}
        style={
          Array [
            Object {
              "color": "#2D3B45",
              "fontSize": 16,
            },
            Array [
              Object {
                "color": "#2D3B45",
                "fontWeight": "600",
              },
            ],
          ]
        }
      >
        Title of the row
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#2D3B45",
              "fontSize": 16,
            },
            Array [
              Object {
                "color": "#8B969E",
                "fontSize": 14,
                "marginTop": 2,
              },
            ],
          ]
        }
        testID="row.undefined-cell-subtitle-lbl"
      >
        Subtitle
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 0,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "flex-end",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignSelf": "center",
                "backgroundColor": "transparent",
                "borderColor": "#C7CDD1",
                "borderRightWidth": 2,
                "borderTopWidth": 2,
                "height": 10,
                "marginLeft": 7,
                "width": 10,
              },
              Object {
                "transform": Array [
                  Object {
                    "rotate": "45deg",
                  },
                ],
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;
